common/base/include/claraparabricks/genomeworks/utils/allocator.hpp

```diff
@@ -114,7 +118,7 @@ class CudaMallocAllocator
     /// \return pointer to allocated memory
     /// \throw device_memory_allocation_exception if allocation was not successful
     pointer allocate(std::size_t n,
-                     const std::vector<cudaStream_t>& streams = {{0}})
+                     const std::vector<cudaStream_t>& streams = {})
```

Comment:
The idea here was not to avoid host memory allocation in `allocate()`. The problem was that when Thrust algorithms use the allocator to allocate temporary memory they do not specify the stream, so the allocation would be associated with the default stream and then when that allocation gets deallocated we would be syncing on the default stream.

I agree about rethinking the way we associate allocations with streams. Another thing I would like to implement is avoiding stream synchronization altogether if the allocation is on the same stream(s) as the old one, like they to in `cub::CachingDeviceAllocator`. In that case once the memory is deallocated there would be no stream synchronization, it would just be marked as unused and an event would probably be created. New allocation would then be able to either take unallocated memory or the memory which is unused and it is associated the the same stream(s).

But I think this should definitely be a separate PR.

Reply:
Ok, I see. True, this should definitely be a separate PR. But here, I also see a quick fix: make a single `cudaStream_t` data member (instead of the full vector) for the default and then construct the vector with this default stream if `allocate()`'s streams argument is empty.